File: app/src/androidTest/java/com/netkaize/subscription/LoginSmokeTest.java
package com.netkaize.subscription;

import android.os.Bundle;
import android.webkit.WebView;

import androidx.test.core.app.ActivityScenario;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.platform.app.InstrumentationRegistry;

import org.json.JSONObject;
import org.junit.Assume;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
public class LoginSmokeTest {
    @Test
    public void productionAccountCanLogInAndReachTheLedger() throws Exception {
        Bundle arguments = InstrumentationRegistry.getArguments();
        String email = arguments.getString("smokeEmail", "");
        String password = arguments.getString("smokePassword", "");
        Assume.assumeTrue("Smoke credentials are required for the production login test",
                !email.isEmpty() && !password.isEmpty());

        try (ActivityScenario<MainActivity> scenario = ActivityScenario.launch(MainActivity.class)) {
            AtomicReference<WebView> webViewRef = new AtomicReference<>();
            scenario.onActivity(activity -> webViewRef.set(activity.getWebViewForTesting()));
            WebView webView = webViewRef.get();

            assertTrue("Login page did not become ready",
                    waitForJavascript(scenario, webView,
                            "Boolean(document.querySelector('#auth-email') && document.querySelector('#auth-submit'))",
                            "true", 30));

            String loginScript = "(() => {"
                    + "const email=document.querySelector('#auth-email');"
                    + "const password=document.querySelector('#auth-password');"
                    + "email.value=" + JSONObject.quote(email) + ";"
                    + "password.value=" + JSONObject.quote(password) + ";"
                    + "email.dispatchEvent(new Event('input',{bubbles:true}));"
                    + "password.dispatchEvent(new Event('input',{bubbles:true}));"
                    + "document.querySelector('#auth-submit').click();"
                    + "return true;"
                    + "})()";
            assertEquals("true", evaluateJavascript(scenario, webView, loginScript));

            assertTrue("The production account did not reach the authenticated ledger",
                    waitForJavascript(scenario, webView,
                            "Boolean(localStorage.getItem('subscription_manager_auth_v1'))"
                                    + " && !document.querySelector('#auth-screen').classList.contains('show')",
                            "true", 30));
        }
    }

    private static boolean waitForJavascript(
            ActivityScenario<MainActivity> scenario,
            WebView webView,
            String script,
            String expected,
            int timeoutSeconds
    ) throws Exception {
        long deadline = System.nanoTime() + TimeUnit.SECONDS.toNanos(timeoutSeconds);
        while (System.nanoTime() < deadline) {
            if (expected.equals(evaluateJavascript(scenario, webView, script))) {
                return true;
            }
            Thread.sleep(500);
        }
        return false;
    }

    private static String evaluateJavascript(
            ActivityScenario<MainActivity> scenario,
            WebView webView,
            String script
    ) throws Exception {
        CountDownLatch latch = new CountDownLatch(1);
        AtomicReference<String> result = new AtomicReference<>();
        scenario.onActivity(activity ->
                webView.evaluateJavascript(script, value -> {
                    result.set(value);
                    latch.countDown();
                })
        );
        assertTrue("JavaScript evaluation timed out", latch.await(10, TimeUnit.SECONDS));
        return result.get();
    }
}

File: app/src/androidTest/java/com/netkaize/subscription/LoginSmokeTest.kt
package com.netkaize.subscription

import androidx.compose.ui.test.hasSetTextAction
import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.compose.ui.test.onNodeWithTag
import androidx.compose.ui.test.onNodeWithText
import androidx.compose.ui.test.performClick
import androidx.compose.ui.test.performTextInput
import androidx.test.platform.app.InstrumentationRegistry
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test

/**
 * Production smoke for the native Compose flow.
 *
 * Credentials are supplied by instrumentation arguments so they never enter the repository or
 * test report. The test fails closed when release credentials are missing.
 */
class LoginSmokeTest {
    @get:Rule
    val composeRule = createAndroidComposeRule<MainActivity>()

    @Test
    fun productionAccountCanLogInAndReadTheLedger() {
        val arguments = InstrumentationRegistry.getArguments()
        val email = arguments.getString("smokeEmail").orEmpty()
        val password = arguments.getString("smokePassword").orEmpty()
        assertTrue(
            "Smoke credentials are required for the production login test",
            email.isNotBlank() && password.isNotBlank(),
        )

        composeRule.waitUntil(timeoutMillis = 20_000) {
            composeRule.onAllNodes(hasSetTextAction()).fetchSemanticsNodes().size >= 2
        }
        val fields = composeRule.onAllNodes(hasSetTextAction())
        fields[0].performTextInput(email)
        fields[1].performTextInput(password)
        composeRule.onNodeWithText("登录并同步").performClick()

        composeRule.waitUntil(timeoutMillis = 30_000) {
            runCatching { composeRule.onNodeWithTag("nav_home").fetchSemanticsNode() }.isSuccess
        }
        composeRule.onNodeWithText("订阅支出概览").assertExists()
        composeRule.onNodeWithTag("nav_subscriptions").performClick()
        composeRule.onNodeWithText("已订阅服务").assertExists()
        composeRule.onNodeWithTag("nav_add").performClick()
        composeRule.onNodeWithText("添加订阅").assertExists()
        composeRule.onNodeWithTag("nav_analysis").performClick()
        composeRule.onNodeWithText("支出分析").assertExists()
        composeRule.onNodeWithTag("nav_profile").performClick()
        composeRule.onNodeWithText("个人中心").assertExists()
        composeRule.onNodeWithText(email).assertExists()
    }
}
